### PR TITLE
Add Ctrl+B shortcut to toggle bold formatting in markdown cells

### LINKED ISSUE #9725: Ctrl-B should bold the selection when in a markdown cell
### Problem

1. Select code in a markdown cell
2. Type `ctrl-b`
3. I'd expect the selection to be fenced in by `**   **` as happens in the github editor

Instead my browser bookmarks bar opens.

### Proposed Solution

`ctrl-b` should take `selection` to `**selection**`
(Ideally it should be invertable by hitting `ctrl-b` again)

### Additional context

This is how the github issues editor works, and I think how many markdown editors work.

### PR DESCRIPTION
**Pull request by MUFFANUJ**
## References
closes #9725

## Code changes
Adds support for toggling bold formatting in markdown cells using the `Ctrl+B` keyboard shortcut (or `Cmd+B` on macOS). When editing a markdown cell, users can now select text and press `Ctrl+B` to wrap it with `**` for bold formatting, matching the behavior of common markdown editors like GitHub. Pressing `Ctrl+B` again on already bold text will remove the bold formatting by unwrapping the `**` markers. The shortcut only works when text is selected in markdown cells, and does not interfere with the existing `Ctrl+B` shortcut for toggling the left sidebar when used elsewhere in the interface.

https://github.com/user-attachments/assets/4928a84d-79be-45e3-8bd4-bab4a5935056
## Backwards-incompatible changes
None

### CODE REVIEW

**Comment by MUFFANUJ** [comment]
@meeseeksdev tag enhancement

**Comment by aaron-seq** [comment]
## Code Review

Thank you for this contribution. I've reviewed the implementation and identified several issues that need to be addressed before this PR can be merged.

### Critical Issues

#### 1. CI/CD Test Failures

The PR has failing tests in the Firefox visual regression tests:
- `test/jupyterlab/notebook-toolbar.test.ts:88:7` - Copy-Paste cell test
- `test/jupyterlab/notebook-toolbar.test.ts:99:7` - Cut cell test  
- `test/jupyterlab/notebook-toolbar.test.ts:108:7` - Paste cell test

These failures appear to be caused by the changes in `galata/src/helpers/notebook.ts` at line 271. The issue is that the new code checks if a markdown cell is focused and presses Escape before clicking toolbar items. This logic interferes with the existing toolbar tests.

**Root cause**: The condition added at line 258-266 escapes from markdown edit mode before toolbar clicks, which breaks the expected behavior in toolbar tests where cells should remain in their current state.

**Recommendation**: The fix should be more targeted. Instead of checking for any focused markdown editor before all toolbar clicks, this escape logic should only apply when necessary (e.g., when the toolbar action specifically requires command mode).

#### 2. Code Logic Issues

**In `packages/codemirror-extension/src/commands.ts`**:

Line 106-110: The bold detection logic is flawed:
```typescript
const isBold =
  selectedText.startsWith('**') && selectedText.endsWith('**');
```

This will incorrectly detect text like `**hello**world**` as bold when only part of it is bold. It will also fail for cases where the selection includes the markers but not the full bold text.

**Better approach**: Check if the selection is immediately surrounded by `**` markers, not if the text itself starts/ends with them.

Line 115-123: Cursor positioning after bold toggle is incorrect. After wrapping text in `**`, the cursor is placed at the end, but it should maintain the selection or position more intuitively.

#### 3. Documentation Issues

The documentation added in `docs/source/user/interface.md` at lines 251-253 has grammatical issues:
- Line 251: Should clarify this only works when text is selected
- Line 253: Awkward phrasing about "text wrapped with `**`"

### Code Quality Suggestions

1. **Test Coverage**: No new tests were added for the `toggleBold` command itself. Unit tests should verify:
   - Wrapping selected text with `**`
   - Unwrapping text that's already bold
   - Handling edge cases (empty selection, partial bold text)

2. **Selector Specificity**: In `packages/codemirror-extension/schema/plugin.json` line 33, the selector `.jp-MarkdownCell .cm-content` is good, but consider if this should also work in the file editor when editing markdown files.

3. **Code Style**: The implementation follows JupyterLab conventions well, but the helper function in `galata/src/helpers/notebook.ts` needs refinement to avoid side effects.

### References

- Original issue: #9725
- Failed CI run: [Visual Regression Tests (firefox)](https://github.com/jupyterlab/jupyterlab/actions/runs/20909276516/job/60091649433?pr=18328)
- JupyterLab Contributing Guidelines: [Contributing to JupyterLab](https://jupyterlab.readthedocs.io/en/latest/developer/contributing.html)

### Action Items

1. Fix the test failures by refining the logic in `galata/src/helpers/notebook.ts`
2. Improve the bold detection logic to handle edge cases correctly
3. Add unit tests for the `toggleBold` command
4. Improve documentation clarity
5. Consider cursor positioning after bold toggle for better UX

Please address these issues and update the PR. I'm happy to provide more detailed guidance on any of these points if needed.

**Comment by MUFFANUJ** [comment]
> #### 1. CI/CD Test Failures
> The PR has failing tests in the Firefox visual regression tests:
> 
> * `test/jupyterlab/notebook-toolbar.test.ts:88:7` - Copy-Paste cell test
> * `test/jupyterlab/notebook-toolbar.test.ts:99:7` - Cut cell test
> * `test/jupyterlab/notebook-toolbar.test.ts:108:7` - Paste cell test
> 
> These failures appear to be caused by the changes in `galata/src/helpers/notebook.ts` at line 271. The issue is that the new code checks if a markdown cell is focused and presses Escape before clicking toolbar items. This logic interferes with the existing toolbar tests.
> 
> **Root cause**: The condition added at line 258-266 escapes from markdown edit mode before toolbar clicks, which breaks the expected behavior in toolba

Hi @aaron-seq thanks for the review but i feel the test failure were not relevent to my changes and have been fixed by https://github.com/jupyterlab/jupyterlab/pull/18335.

**Comment by MUFFANUJ** [comment]
> * Line 251: Should clarify this only works when text is selected

i feel its clearly mentioned in the very next line `Select text in a markdown cell and press {kbd}`Ctrl+B` to wrap it with `**` for bold formatting`.


> * Line 253: Awkward phrasing about "text wrapped with `**`"

hmm i think its was just providing the clarification about the bold text formatting?

**Comment by MUFFANUJ** [comment]
> **Test Coverage**: No new tests were added for the `toggleBold` command itself. Unit tests should verify:

i will be providing a UI test for it thanks :)

File: packages/codemirror-extension/schema/plugin.json [hunk 1]
```diff
@@ -26,6 +26,12 @@
       "command": "codemirror:select-next-occurrence",
       "keys": ["Accel Shift D"],
       "selector": ".cm-content"
+    },
+    {
+      "command": "codemirror:toggle-bold",
+      "keys": ["Accel B"],
+      "selector": ".jp-MarkdownCell .cm-content",
```
Comment (krassowski) [inline packages/codemirror-extension/schema/plugin.json]: It would be great to allow this for any Markdown editor, not only in Markdown cells.

File: packages/codemirror-extension/schema/plugin.json [hunk 2]
```diff
@@ -26,6 +26,12 @@
       "command": "codemirror:select-next-occurrence",
       "keys": ["Accel Shift D"],
       "selector": ".cm-content"
+    },
+    {
+      "command": "codemirror:toggle-bold",
+      "keys": ["Accel B"],
```
Comment (krassowski) [inline packages/codemirror-extension/schema/plugin.json:32]: This kind of conflicts with show/hide file browser, but since it is still accessible in the command mode I would say we should give it a go in 4.6.0 pre-releases (should there be a push back we can look for an alternative default shortcut).

File: packages/codemirror-extension/schema/plugin.json (hunk 1 above)
Comment (krassowski) [reply packages/codemirror-extension/schema/plugin.json]: I did not test this, but it feels like it should work:

```suggestion
      "selector": ".cm-content[data-language='markdown']",
```

File: packages/codemirror-extension/src/commands.ts [hunk 3]
```diff
@@ -231,6 +272,27 @@ export const commandsPlugin: JupyterFrontEndPlugin<void> = {
...
+        }
+      },
+      execute: () => {
+        const view = findEditorView();
+        if (!view) {
+          return;
+        }
+        toggleBold(view);
+      },
+      isEnabled
```
Comment (krassowski) [inline packages/codemirror-extension/src/commands.ts:315]: Maybe worth adding a check if editor view selection is using markdown mode? Not sure.

**Comment by krassowski** [comment]
> Better approach: Check if the selection is immediately surrounded by \*\* markers, not if the text itself starts/ends with them.

I think we should do both. Basically whether there are \*\* on the inside or outside of the selection - this is what GitHub does.

Also, the selection should persist after operation - this is what GitHub does.

File: packages/codemirror-extension/schema/plugin.json (hunk 1 above)
Comment (MUFFANUJ) [reply packages/codemirror-extension/schema/plugin.json]: sure

File: packages/codemirror-extension/schema/plugin.json (hunk 2 above)
Comment (MUFFANUJ) [reply packages/codemirror-extension/schema/plugin.json:32]: sure thing

File: packages/codemirror-extension/src/commands.ts (hunk 3 above)
Comment (MUFFANUJ) [reply packages/codemirror-extension/src/commands.ts:315]: not sure if we should add this as shortcut selector already narrows to Markdown `.cm-content[data-language='markdown']`

File: packages/codemirror-extension/src/commands.ts [hunk 4]
```diff
@@ -38,6 +39,7 @@ namespace CommandIDs {
   export const toggleComment = 'codemirror:toggle-comment';
   export const selectNextOccurrence = 'codemirror:select-next-occurrence';
   export const toggleTabFocusMode = 'codemirror:toggle-tab-focus-mode';
+  export const toggleBold = 'codemirror:toggle-bold';
```
Comment (krassowski) [inline packages/codemirror-extension/src/commands.ts]: I wonder if we should be more specific with naming here:

```suggestion
  export const toggleBold = 'codemirror:markdown-toggle-bold';
```

(of course would need to be changed in the other places too)

Comment (MUFFANUJ) [reply packages/codemirror-extension/src/commands.ts]: You are right, `codemirror:markdown-toggle-bold` feels more better than just `codemirror:toggle-bold`

**Comment by jasongrout** [comment]
+1 to having this shortcut. Thanks!
